Source/WebKit/Shared/Cocoa/CoreIPCPlistObject.mm

```diff
@@ -0,0 +1,120 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "CoreIPCPlistObject.h"
+
+#if PLATFORM(COCOA)
+
+#import "ArgumentCodersCocoa.h"
+#import "CoreIPCData.h"
+#import "CoreIPCDate.h"
+#import "CoreIPCNumber.h"
+#import "CoreIPCPlistArray.h"
+#import "CoreIPCPlistDictionary.h"
+#import "CoreIPCString.h"
+#import "GeneratedWebKitSecureCoding.h"
+#import <wtf/cocoa/TypeCastsCocoa.h>
+
+namespace WebKit {
+
+bool CoreIPCPlistObject::isPlistType(id value)
+{
+    if ([value isKindOfClass:[NSString class]]
+        || [value isKindOfClass:[NSArray class]]
+        || [value isKindOfClass:[NSData class]]
+        || [value isKindOfClass:[NSNumber class]]
+        || [value isKindOfClass:[NSDate class]]
+        || [value isKindOfClass:[NSDictionary class]])
+        return true;
+    return false;
+}
+
+static PlistValue valueFromID(id object)
+{
+    if (!object)
+        return nullptr;
+
+    switch (IPC::typeFromObject(object)) {
+    case IPC::NSType::Array:
+        return CoreIPCPlistArray((NSArray *)object);
+    case IPC::NSType::Data:
+        return CoreIPCData((NSData *)object);
+    case IPC::NSType::Date:
+        return CoreIPCDate(bridge_cast((NSDate *)object));
+    case IPC::NSType::Dictionary:
+        return CoreIPCPlistDictionary((NSDictionary *)object);
+    case IPC::NSType::Number:
+        return CoreIPCNumber(bridge_cast((NSNumber *)object));
+    case IPC::NSType::String:
+        return CoreIPCString((NSString *)object);
+    default:
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+}
+
+CoreIPCPlistObject::CoreIPCPlistObject(id object)
+    : m_value(makeUniqueRefWithoutFastMallocCheck<PlistValue>(valueFromID(object)))
+{
+}
+
+CoreIPCPlistObject::CoreIPCPlistObject(UniqueRef<PlistValue>&& value)
+    : m_value(WTFMove(value))
+{
+}
+
+RetainPtr<id> CoreIPCPlistObject::toID() const
+{
+    RetainPtr<id> result;
+
+    WTF::switchOn(*m_value, [&](auto& object) {
+        result = object.toID();
+    }, [](std::nullptr_t) {
+        // result should be nil, which is the default value initialized above.
```

Comment:
This would be slightly nicer like this:

return WTF::switchOn(*m_value, [&](auto& object) {
    return object.toID();
}, [](std::nullptr_t) {
    return RetainPtr<id> { };
});